addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/handler/HttpThingHandler.java

```diff
@@ -0,0 +1,138 @@
+
+    public HttpThingHandler(final Thing thing,
+                            final HttpClient httpClient,
+                            final ItemChannelLinkRegistry itemChannelLinkRegistry)
+    {
+        super(thing);
+        this.httpClient = httpClient;
+        this.itemChannelLinkRegistry = itemChannelLinkRegistry;
+    }
+
+    @Override
+    public void initialize() {
+        try {
+            for (final Channel channel : getThing().getChannels()) {
+                final ChannelTypeUID channelTypeUID = channel.getChannelTypeUID();
+                if (channelTypeUID != null) {
+                    final int maxHttpResponseBodyLen;
+                    if (CHANNEL_TYPE_ID_IMAGE.equals(channelTypeUID.getId())) {
+                        maxHttpResponseBodyLen = MAX_IMAGE_RESPONSE_BODY_LEN;
+                    } else {
+                        maxHttpResponseBodyLen = MAX_RESPONSE_BODY_LEN;
+                    }
+
+                    final HttpChannelConfig config = channel.getConfiguration().as(HttpChannelConfig.class);
+                    final HttpChannelState channelState = new HttpChannelState(
+                            channel.getUID(),
+                            channelTypeUID,
+                            this.httpClient,
+                            maxHttpResponseBodyLen,
+                            config.getStateRequest(bundleContext),
+                            scheduler,
+                            config.getCommandRequest(bundleContext),
+                            this::updateState,
+                            this::communicationsError
+                    );
+                    channels.put(channel.getUID(), channelState);
+                }
+            }
+            updateStatus(ThingStatus.ONLINE);
+        } catch (final IllegalArgumentException e) {
+            disposeChannels();
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+        }
+    }
+
+    @Override
+    public void handleCommand(final ChannelUID channelUID, final Command command) {
+        if (channels.containsKey(channelUID)) {
+            channels.get(channelUID).handleCommand(command);
+        }
+    }
+
+    @Override
+    protected void updateStatus(final ThingStatus status, final ThingStatusDetail statusDetail, @Nullable final String description) {
+        final ThingStatusInfo curStatusInfo = getThing().getStatusInfo();
+        if (!Objects.equals(status, curStatusInfo.getStatus()) || !Objects.equals(statusDetail, curStatusInfo.getStatusDetail()) || !Objects.equals(description, curStatusInfo.getDescription())) {
+            super.updateStatus(status, statusDetail, description);
+        }
+    }
+
+    @Override
+    protected void updateState(final ChannelUID channelUID, final State state) {
+        final boolean needsUpdate = this.itemChannelLinkRegistry.getLinkedItems(channelUID).stream()
+                .anyMatch(item -> !state.equals(item.getState()));
```

Comment:
That is now more efficient that the solution before. But how about using
```BaseThingHandler.isLinked```? You could remove `itemChannelLinkRegistry` completely.

Reply:
So what I'm trying to accomplish here is to avoid `ItemStateEvent`s when the item's state hasn't actually changed.  Since we have to poll the HTTP URL for changes, each poll might fetch the same data as last time.  `isLinked` will of course tell me if the channel is linked to some item, but won't tell me the current state of that item and if the new state we've just fetched is actually different.

Initially I thought that the state can't change in a way such that we wouldn't know about it -- could just cache the last `Command` sent to the Thing, and then when we fetch state, compare it to that last command.  But, for example, for `INCREASE` and `DECREASE` commands, we don't know what the new state is after the HTTP server has processed the command, so on the next state fetch we don't know if the old state is valid or not.  We also don't necessarily know what to do if the command URL returns a failure status code: revert to the old state?  Change to some other unknown state?

I was thinking that the "protocol" I've made up between the binding and the HTTP servers might require that the Command URL return the new state of the device on success, but I expect that might be too much of a burden on existing HTTP services that might just not do that, and might not be modifiable.

But I guess there's a way to get most of the way there:

1. Cache the returned state after each request to the State URL.  Don't update state if the new state is the same as the last returned state.
2. Cache the last command sent after, if possible, converting it to an expected state.  If the last command sent isn't convertible to state (e.g. `INCREASE`), just clear out the cached state since it's now unknown.
3. After sending a command, immediately also fetch state (ignoring the usual refresh interval) in order to ensure our view of the world is correct.  (Maybe only do this for when we can't convert a command into a state.)

I'll think about this a bit more... will address other comments and come back to this later.

Reply:
Most bindings just update channels with the same value. And the framework actually keeps an eye on the value, because you either get item changed or item updated events on the event bus. I would just push the value to the channel and use isLinked to not do any http request if not necessary.

Reply:
Ah, didn't realize there was a difference between changed and updated.  Removing this check will simplify things quite a bit!  I'll also remove the change in `updateStatus()` for the Thing status updates.